🎉 New Source: Exact

Jeroendevr opened the pull request:
## What
Implements source using Python CDK for [Exact Online](https://www.exact.com)

However I need some guidance to the right documentation, as right now the testing or buidling of the connector fails. 

### Related 
- This is an updated attempt to incorporate the new Airbyte version with the PR of https://github.com/airbytehq/airbyte/pull/20480
- Connector request see discussion https://github.com/airbytehq/airbyte/discussions/33178

## Reading order
The code of which I am unsure `airbyte-integrations/connectors/source-exact/pyproject.toml`


## Review guide and questions
1. It seems Poetry had an upgrade, does it matter for the `pyproject.toml` or building of the connector in general if the Poetry version is 1.8.3 or 2.1.3?
2. Running `airbyte-ci connectors --name=source-exact test` gives me the following output:
```sh
pipelines: An error occurred while running the test pipeline
╭────────────────────── Traceback (most recent call last) ───────────────────────╮
│ in test:175                                                                    │
│                                                                                │
│ in run_connectors_pipelines:111                                                │
│                                                                                │
│ in __aexit__:772                                                               │
╰────────────────────────────────────────────────────────────────────────────────╯
ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
pipelines: Dagger Command test failed             
╭─────────────── Traceback (most recent call last) ────────────────╮
│ in invoke:52                                                     │
╰──────────────────────────────────────────────────────────────────╯
DaggerError: Dagger Command test failed.
pipelines: Dagger logs saved to ~/airbyte/airbyte-ci/connectors/pipelines/pipe
line_reports/airbyte-ci/connectors/test/manual/master/1749725060/bf9
a12dc973a87b1879c97b45d626097e4f38bfb/dagger.log
```

But the `dagger.log` is empty, so how to debug this exact (no pun intended) step?
3. Running  `airbyte-ci connectors --name=source-exact build`
```sh
Failed to build connector image for platform linux/arm64: resolve: process "poetry check" did not complete successfully: exit code: 1
Stdout:

Stderr:
Error: pyproject.toml changed significantly since poetry.lock was last generated. Run `poetry lock [--no-update]` to fix the lock file.
```
- This is a step indicating that during build Poetry uses version `1.x.x` as `poetry lock --no-update` is a v1 command. 
- Running `poetry check` locally yields some warnings however no errors. How to debug this step?
 - Mainly about the difference layout of `[project]` vs `[tool.poetry]` in the `pyproject.toml` file.  

## Can this PR be safely reverted and rolled back?
<!--
* If unsure, leave it blank.
-->
- [ ] YES 💚
- [ ] NO ❌

Jeroendevr committed:
able to run spec command

Jeroendevr committed:
svae before test with manifest adjustments

Jeroendevr committed:
restart from abstract source class

Jeroendevr committed:
working spec

Jeroendevr committed:
commit before running discover

Jeroendevr committed:
added schema for account classifications

Jeroendevr committed:
commit before configuring authentication

Jeroendevr committed:
keuze voor Oauth2authenticator aangezien refresh token 30 dagen geldig id

Jeroendevr committed:
use of correct endpoint

Jeroendevr committed:
remove poetry references and updates with use of UV

Jeroendevr committed:
Refreshtoken is toch single use, Access token niet, daarom toch gebruiken maken van SingleUseRefreshTokenOauth2Authenticator

Jeroendevr committed:
successfull check with token refresh

Jeroendevr committed:
Limitation of multiple divisions

Jeroendevr committed:
Comment out tests which require different setup

Jeroendevr committed:
explicit override of config param during test

Jeroendevr committed:
State setter and getter properties

Jeroendevr committed:
Pin repo to python version as used by the base image.
Downgraded python version to satisy build

Jeroendevr committed:
set base image to 4.0.0 with python to 3.11

Jeroendevr committed:
Add lock file to check workings of poetry

Jeroendevr committed:
update poetry tool section

Jeroendevr committed:
Build works but gives error

Jeroendevr committed:
defined the use of poe in pyproject.toml

Jeroendevr committed:
moved unit tests to connector root

Jeroendevr committed:
passed unit tests using poe

Jeroendevr committed:
created a pyproject.toml with poetry v1

Jeroendevr committed:
updated source-exact script

Jeroendevr committed:
added shared task definition, and include the package

Jeroendevr committed:
data.include should be an array

Jeroendevr committed:
lock file

Jeroendevr committed:
added pendulum

Jeroendevr committed:
ruff format on main py files

Jeroendevr committed:
ruff

Jeroendevr committed:
ruff formatted and checked with prettier

Jeroendevr committed:
run prettier on source

Jeroendevr committed:
license headers

Jeroendevr committed:
docker repository added

Jeroendevr committed:
changed to toolpoetry.scripts setting in pyproject.toml

Jeroendevr committed:
added definition ID

Jeroendevr committed:
added documentation and icon

Jeroendevr committed:
lower case and cdk tag

Jeroendevr committed:
pypi must have a name

Jeroendevr committed:
moved spec to connector folder, passes airbyte-cdk image test command

Jeroendevr committed:
ruff format

Jeroendevr committed:
return the connector spec

Jeroendevr committed:
update docker image tag

Jeroendevr committed:
update patch version 0.2.3

Jeroendevr committed:
wip integration_test.py

Jeroendevr committed:
new schema and updated a stream slices method from the base connecto

Jeroendevr committed:
ruff formatting

Jeroendevr committed:
add stream slice and authenticator

Jeroendevr committed:
remove old comment

Jeroendevr committed:
utils function for easily working with local secrets

Jeroendevr committed:
added new schemas

Jeroendevr committed:
we dont set prod on airbyte instance therfore change is dev

Jeroendevr committed:
Added new streams like project_invoice_terms.json

Jeroendevr committed:
added stream syncPayrollEmployments

Jeroendevr committed:
added stream hrm_depart

Jeroendevr committed:
added stream sync payroll

Jeroendevr committed:
New stream sync_payroll_employees.json

Jeroendevr committed:
added stream schema for payroll_active_employments.json

Jeroendevr committed:
Added stream openingbalance_current_year.json

Jeroendevr committed:
add stream financial_recievables_list.json

Jeroendevr committed:
Added stream budget_budgets.json

Jeroendevr committed:
added new stream financial_gl_account_classification_mappings.json

Jeroendevr committed:
Added stream financial_reporting_balance.json

Jeroendevr committed:
Update to cursor field of FinancialReceivablesList

Jeroendevr committed:
for openingbalance_current_year_after_entry.json there is no cursor

Jeroendevr committed:
incorrect endpoints, dont start endpoint with '/'

Jeroendevr committed:
incorrect schema for sync_payroll_employees.json

Jeroendevr committed:
Updated schema of project_invoice_terms.json and budget_budgets.json

Jeroendevr committed:
Update to project_invoice_terms.json schema

Jeroendevr committed:
check connection changed to not access a stream endpoint but simply retrieve current division

Jeroendevr committed:
streams without specific cursor set primary key to cursor as filter is used in Exact onlines docs

Jeroendevr committed:
Set integration test to use projects endpoint and the dev variable.

Jeroendevr committed:
ruff formatting

Jeroendevr committed:
fix unit tests params

Jeroendevr committed:
explicit stream tests to verify endpoint

cgardens commented:
@Jeroendevr can you try building this with connector builder. I think it'll help avoid some of the current issues you're running into.

Jeroendevr commented:
_Connector Builder_
Yes would love to develop using Connector Builder. I have developed some connectors already. However in my previous attempts I did not managed to get it working. The debug messages weren't clear to me. Would you or someone else from Airbyte be able to provide a little bit of support?

_meanwhile_
I made some progress and got the connector working (in production 🙂). Shall I update my opening post?  In the docs and readme, I've provided the up to date instructions.

cgardens commented:
/run-connector-tests

cgardens commented:
/format-fix

Jeroendevr commented:
Hi @cgardens yes will do, let me check the coming month to make some time.

DanyloGL commented:
@Jeroendevr hi, this is connector's Initial release so it should be `0.1.0` version.
Could you please also fix the following issues:
`source-exact - ❌ Failed - Connectors must have a changelog entry for each version: Connectors must have a changelog entry for each version: changelog entry for version 0.2.20 is missing in the documentation.`

`source-exact - ❌ Failed - Python connectors must have a CDK tag in metadata: Expected CDK tag 'cdk:low-code' in the metadata.yaml file, but found 'cdk:python'.`

Thanks!

Jeroendevr committed:
New integration test for connection checking. Updated changelog and project files to set version to 0.1.0

Jeroendevr committed:
low-code manifest removal

Jeroendevr commented:
Hi @DanyloGL Thanks for pointing out. I have updated the changelog to 0.1.0. 
Regarding the `cdk` tag. Perhaps the CI was confused because of the existence of a `manifest.yaml` file. I have removed the file. The tag `cdk-python` should be correct according to the [docs](https://docs.airbyte.com/platform/contributing-to-airbyte/resources/qa-checks#python-connectors-must-have-a-cdk-tag-in-metadata)

> Python connectors must have a CDK tag in their metadata. It must be set in the tags field in metadata.yaml. The values can be cdk:low-code, cdk:python, or cdk:file.

Jeroendevr committed:
prettier formatting of schema

Jeroendevr committed:
Add stream logistics items

Jeroendevr committed:
Add new stream

Jeroendevr committed:
adding two streams and formatting

Jeroendevr committed:
add new streams to exact source

Jeroendevr committed:
removed unnecessary stream fields

Jeroendevr committed:
Removal of unesecary  fields a

Jeroendevr committed:
new schema for internal rates

Jeroendevr committed:
ruff formatting

Jeroendevr committed:
update changelog with internal rates

Jeroendevr committed:
Merge branch 'airbytehq:master' into master

Jeroendevr committed:
fix on InternalRate stream

Jeroendevr committed:
Merge branch 'airbytehq:master' into master

Jeroendevr committed:
Merge branch 'airbytehq:master' into master
